tsdb/chunks/head_chunks.go

```diff
@@ -0,0 +1,753 @@
+		if err := cdm.cut(mint); err != nil {
+			return 0, err
+		}
+		cdm.curFileMaxt = 0
+	}
+
+	// seriesRef(8) + mint(8) + maxt(8) + encoding(1) + max uvarint size (5) + the bytes is
+	// the total size to be written.
+	// if len(chk.Bytes()) >= writeBufferSize-30, it means that chunk >= the buffer size;
+	// so no need to flush here, as we have to flush at the end (to not keep partial chunks in buffer).
+	if len(chk.Bytes()) < writeBufferSize-30 && cdm.wbuf.Available() < 8+8+8+1+5+len(chk.Bytes()) {
+		if err := cdm.flushBuffer(); err != nil {
+			return 0, err
+		}
+	}
+
+	cdm.crc32.Reset()
+	binary.BigEndian.PutUint64(cdm.writeBuf[:], seriesRef)
+	if err := cdm.writeWithCRC32(cdm.writeBuf[:SeriesRefSize]); err != nil {
+		return 0, err
+	}
+
+	binary.BigEndian.PutUint64(cdm.writeBuf[:], uint64(mint))
+	if err := cdm.writeWithCRC32(cdm.writeBuf[:MintMaxtSize]); err != nil {
+		return 0, err
+	}
+
+	binary.BigEndian.PutUint64(cdm.writeBuf[:], uint64(maxt))
+	if err := cdm.writeWithCRC32(cdm.writeBuf[:MintMaxtSize]); err != nil {
+		return 0, err
+	}
+
+	// The reference is set to the head chunk file index and the offset where
+	// the data starts for this chunk.
+	//
+	// The upper 4 bytes are for the head chunk file index and
+	// the lower 4 bytes are for the head chunk file offset where to start reading this chunk.
+	chunkRef = cdm.chunkRef(uint64(cdm.curFileSequence), uint64(cdm.n))
+
+	cdm.writeBuf[0] = byte(chk.Encoding())
+	if err := cdm.writeWithCRC32(cdm.writeBuf[:1]); err != nil {
+		return 0, err
+	}
+
+	n := binary.PutUvarint(cdm.writeBuf[:], uint64(len(chk.Bytes())))
+	if err := cdm.writeWithCRC32(cdm.writeBuf[:n]); err != nil {
+		return 0, err
+	}
+
+	if err := cdm.writeWithCRC32(chk.Bytes()); err != nil {
+		return 0, err
+	}
+
+	if err := cdm.write(cdm.crc32.Sum(cdm.writeBuf[:0])); err != nil {
+		return 0, err
+	}
+
+	if maxt > cdm.curFileMaxt {
+		cdm.curFileMaxt = maxt
+	}
+
+	if mint < cdm.curFileMint {
+		cdm.curFileMint = mint
+		// As this wont happen a whole lot of time, we don't wait
+		// for a new head chunk file to be cut to write it off.
+		binary.BigEndian.PutUint64(cdm.writeBuf[:], uint64(mint))
+		if _, err := cdm.curFile.WriteAt(cdm.writeBuf[:MintMaxtSize], HeaderMintOffset); err != nil {
+			return 0, err
+		}
+	}
+
+	cdm.inBufferChunksLocks[chunkRef%inBufferShards].Lock()
+	cdm.inBufferChunks[chunkRef%inBufferShards][chunkRef] = chk
+	cdm.inBufferChunksLocks[chunkRef%inBufferShards].Unlock()
+
+	if len(chk.Bytes()) >= writeBufferSize-30 {
+		if err := cdm.flushBuffer(); err != nil {
+			return 0, err
+		}
+	}
+
+	return chunkRef, nil
+}
+
+func (cdm *ChunkDiskMapper) chunkRef(seq, offset uint64) (chunkRef uint64) {
+	return (seq << 32) | offset
+}
+
+func (cdm *ChunkDiskMapper) shouldCutNewFile(chunkLength int, maxt int64) bool {
+	return cdm.n == 0 || // First head chunk file.
+		(maxt-cdm.curFileMint > cdm.maxFileMs && cdm.n > HeadChunkFileHeaderSize) || // Time duration reached for the existing file.
```

Comment:
Time: so that we can delete the files kind of deterministically after some time. File deletion is based on time.
Size: OS has limits on how big of a file we can mem-map, so we limit at that too.

Reply:
Hm, can we mention that in comment. I think it's unlikely we get out OS limit, but sure (:

Reply:
`MaxInt32` (the max size for file now) it just ~1GiB :) so it is easily possible to cross it in high load environments. I will add a comment, and maybe also explicitly have it as 1 GiB instead of MaxInt32.